selfdrive/locationd/magd.py

```diff
@@ -0,0 +1,221 @@
+        if np.all(np.abs(self.raw_vals - self.past_raw_vals) < 2.0) and msg.magneticUncalibrated.status:
+          self.filtered_vals = np.array([f.update(v) for f, v in zip(self.filtered_magnetic, self.raw_vals, strict=True)])
+        if self.vego > MIN_VEL and self.yaw_valid:
+          self.point_buckets.add_point(self.filtered_vals[0], self.filtered_vals[2], self.yaw)
+        self.past_raw_vals = self.raw_vals
+
+  def get_msg(self, valid: bool = True, with_points: bool = False) -> log.Event:
+    msg = messaging.new_message('magnetometerCalibration')
+    msg.valid = valid
+    magnetometerCalibration = msg.magnetometerCalibration
+    magnetometerCalibration.version = VERSION
+
+    bearing, bearingValid = 0.0, False
+    if self.calibrated:
+      bearing = self.get_calibrated_bearing(self.filtered_vals[0], self.filtered_vals[2], self.calibrationParams)
+      bearingValid = True
+
+    magnetometerCalibration.calibrated = self.calibrated
+    magnetometerCalibration.calibrationParams = self.calibrationParams.tolist()
+    magnetometerCalibration.bearing = float(bearing)
+    magnetometerCalibration.bearingValid = bearingValid
+    magnetometerCalibration.totalBucketPoints = len(self.point_buckets)
+    if with_points:
+      magnetometerCalibration.points = self.point_buckets.get_points().tolist()
+    return msg
+
+  def compute_calibration_params(self) -> None:
+    if self.point_buckets.is_valid():
+      calibration_params = self.estimate_calibration_params()
+      if any(np.isnan(calibration_params)):
+        self.calibrationParams = CALIB_DEFAULTS
+        self.calibrated = False
+      else:
+        self.calibrationParams = calibration_params
+        self.calibrated = True
+
+
+def main(sm: Optional[messaging.SubMaster] = None, pm: Optional[messaging.PubMaster] = None):
+  config_realtime_process([0, 1, 2, 3], 5)
+
+  if sm is None:
+    sm = messaging.SubMaster(['magnetometer', 'carState', 'liveLocationKalman'], poll=['liveLocationKalman'])
+
+  if pm is None:
+    pm = messaging.PubMaster(['magnetometerCalibration'])
+
+  params = Params()
+  with car.CarParams.from_bytes(params.get("CarParams", block=True)) as CP:
+    calibrator = MagCalibrator(CP)
+
+  def cache_params(sig, frame):
+    signal.signal(sig, signal.SIG_DFL)
+    cloudlog.warning("caching mag calib params")
+
+    params = Params()
+    params.put("MagnetometerCarParams", CP.as_builder().to_bytes())
+
+    msg = calibrator.get_msg(with_points=True)
+    params.put("MagnetometerCalibration", msg.to_bytes())
+
+    sys.exit(0)
+  if "REPLAY" not in os.environ:
+    signal.signal(signal.SIGINT, cache_params)
+
+  while True:
+    sm.update()
+    if sm.all_checks():
+      for which in sm.updated.keys():
+        if sm.updated[which]:
+          calibrator.handle_log(which, sm[which])
+
+    if sm.updated['magnetometer']:
+      pm.send('magnetometerCalibration', calibrator.get_msg(valid=sm.all_checks()))
+
+    # 1Hz
+    if sm.frame % 20 == 0:
```

Comment:
looks like this is missing a check on `sm.all_checks` and it can be running with stale (or invalid) data

Reply:
Computing the calibration params is not affected by current sm checks state. It uses values in the point buckets if they are valid.